### PR TITLE
M0-2-1: make into split and shared query

### LINKED ISSUE #119: `M0-2-1`: Consider array copies with overlapping data
### Affected rules
 - `M0-2-1`

### Description

The query as currently written only considers overlapping as caused by unions. We should also consider whether overlapping arrays are covered by the same rule.

### Example

```cpp
#include <cstring>
int16_t a[20];
void f2(void) {
  std::memcpy(&a[0], &a[1], 10u * sizeof(a[0]));  // Non-compliant
  std::memmove(&a[0], &a[1], 10u * sizeof(a[0])); // Compliant 
  std::memcpy(&a[1], &a[0], 10u * sizeof(a[0]));  // Non-compliant
  std::memmove(&a[1], &a[0], 10u * sizeof(a[0])); // Compliant 
}
```

### PR DESCRIPTION
**Pull request by knewbury01**
## Description

additionally moved some libraries to allow for use in shared query

Fixes #119 

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - M0-2-1
     - EXP43-C

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [ ] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [ ] Have all the relevant rule package description files been checked in?
 - [ ] Have you verified that the metadata properties of each new query is set appropriately?
 - [ ] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [ ] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [ ] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [ ] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### Reviewer 

 - [ ] Have all the relevant rule package description files been checked in?
 - [ ] Have you verified that the metadata properties of each new query is set appropriately?
 - [ ] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [ ] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [ ] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [ ] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: rule_packages/cpp/Representation.json [hunk 1]
```diff
@@ -53,6 +53,18 @@
           "tags": [
             "correctness"
           ]
+        },
+        {
+          "description": "Passing an aliased pointer to a conceptually restrict-qualified parameter is undefined behavior.",
```
Comment (nicolaswill) [inline rule_packages/cpp/Representation.json:58]: Could you please clarify what you mean by "conceptually"?

File: rule_packages/c/Pointers3.json [hunk 2]
```diff
@@ -72,6 +72,7 @@
           "precision": "medium",
           "severity": "error",
           "short_name": "DoNotPassAliasedPointerToRestrictQualifiedParam",
+          "shared_implementation_short_name": "DoNotPassAliasedPointerToRestrictQualifiedParam_Shared",
```
Comment (nicolaswill) [inline rule_packages/c/Pointers3.json:75]: Our other shared queries typically use "Shared" as a suffix as opposed to "_Shared," if there is any suffix at all.

File: rule_packages/cpp/Representation.json [hunk 3]
```diff
@@ -53,6 +53,18 @@
...
+          "kind": "problem",
+          "name": "Do not pass aliased pointers as parameters of functions where it is undefined behaviour for those pointers to overlap",
+          "precision": "medium",
+          "severity": "error",
+          "short_name": "DoNotPassAliasedPointerToRestrictQualifiedParam",
+          "shared_implementation_short_name": "DoNotPassAliasedPointerToRestrictQualifiedParam_Shared",
```
Comment (nicolaswill) [inline rule_packages/cpp/Representation.json:64]: See other suffix comment

**Review by nicolaswill — CHANGES_REQUESTED** [review]
LGTM other than some naming and documentation points

File: rule_packages/cpp/Representation.json (hunk 1 above)
Comment (knewbury01) [reply rule_packages/cpp/Representation.json:58]: was mostly referring to the fact that the `restrict` keyword comes from C (not C++) so technically this rule means something slightly different in C++ but checks the same concept

ie - in the C version of the [memcpy description](https://en.cppreference.com/w/c/string/byte/memcpy) - overlapping arrays are the problem (but using those violates the restrict contract)

in the C++ [memcpy description](https://en.cppreference.com/w/cpp/string/byte/memcpy) there is no restrict contract broken, but it is just that the arrays overlap

I think I can just reword this to make it clearer for the C++ version that its really only about overlapping arrays

**Review by nicolaswill — APPROVED** [review]
LGTM!
